KAFKA-10818: Skip conversion to `Struct` when serializing generated requests/responses

ijuma opened the pull request:
Generated request/response classes have code to serialize/deserialize directly to
`ByteBuffer` so the intermediate conversion to `Struct` can be skipped for them.
We have recently completed the transition to generated request/response classes,
so we can also remove the `Struct` based fallbacks.

Additional noteworthy changes:
* `AbstractRequest.parseRequest` has a more efficient computation of request size that
relies on the received buffer instead of the parsed `Struct`.
* Use `SendBuilder` for `AbstractRequest/Response` `toSend`, made the superclass
implementation final and removed the overrides that are no longer necessary.
* Removed request/response constructors that assume latest version as they are unsafe
outside of tests.
* Removed redundant version fields in requests/responses.
* Removed unnecessary work in `OffsetFetchResponse`'s constructor when version >= 2.
* Made `AbstractResponse.throttleTimeMs()` abstract.
* Using `toSend` in `SaslClientAuthenticator` instead of `serialize`.
* Various changes in Request/Response classes to make them more consistent and to
rely on the Data classes as much as possible when it comes to their state.
* Remove the version argument from `AbstractResponse.toString`.
* Fix `getErrorResponse` for `ProduceRequest` and `DescribeClientQuotasRequest` to
use `ApiError` which processes the error message sent back to the clients. This was
uncovered by an accidental fix to a `RequestResponseTest` test (it was calling
`AbstractResponse.toString` instead of `AbstractResponse.toString(short)`).

Rely on existing protocol tests to ensure this refactoring does not change 
observed behavior (aside from improved performance).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

ijuma committed:
Skip redundant `toStruct()` step when serializing requests that. rely on the auto generated protocol

ijuma committed:
Skip redundant `toStruct` when serializing responses that rely on auto generated protocol

ijuma commented:
Started a system tests run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3052/

ijuma commented:
retest this please

ijuma committed:
Add ignorable to TopicAuthorizedOperation and ClusterAuthorizedOperations

ijuma committed:
More ignorable

ijuma committed:
Add missing whitespace

ijuma commented:
Scala 2.13 build passed, Scala 2.12 timed out, 2.11 failed due to H40 slave being offline.

ijuma commented:
retest this please

ijuma committed:
Move `message()` to `AbstractRequestResponse` and document it

ijuma committed:
Use `toSend` instead of `serializeWithHeader` in `SaslClientAuthenticator`.

ijuma committed:
Don't rely on `Struct` directly when parsing requests in the broker

ijuma committed:
Avoid intermediate structs when deserializing in the broker

ijuma committed:
Avoid intermediate Struct when deserializing responses

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -697,21 +695,20 @@ public Node leastLoadedNode(long now) {
-        if (throttleTimeSensor != null && responseBody.hasField(CommonFields.THROTTLE_TIME_MS))
-            throttleTimeSensor.record(responseBody.get(CommonFields.THROTTLE_TIME_MS), now);
-        return responseBody;
+        if (throttleTimeSensor != null && response.throttleTimeMs() != AbstractResponse.DEFAULT_THROTTLE_TIME)
```
Need to verify that this condition is equivalent to the previous behavior of checking if the field exists.

ijuma committed:
Add back removed tests, modify them to test parsing from buffers instead of structs

ijuma commented:
retest this please

ijuma committed:
Avoid `toStruct` for request/response header if the request/response uses the auto generated protocol

ijuma committed:
Add `toString` to `ResponseHeader`

ijuma committed:
Use diamond

ijuma committed:
Remove unnecessary boxing/unboxing

ijuma committed:
Use assertThrows to simplify code

ijuma committed:
Hacky test fix

ijuma committed:
Add comment

ijuma commented:
retest this please

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk:
  KAFKA-8932; Add tag for CreateTopicsResponse.TopicConfigErrorCode (KIP-525) (#7464)
  KAFKA-8944: Fixed KTable compiler warning. (#7393)
  KAFKA-8964: Rename tag client-id for thread-level metrics and below (#7429)
  MINOR: remove unused imports in Streams system tests (#7468)
  KAFKA-7190; Retain producer state until transactionalIdExpiration time passes (#7388)
  KAFKA-8983; AdminClient deleteRecords should not fail all partitions unnecessarily (#7449)
  MINOR: Modified Exception handling for KIP-470 (#7461)
  KAFKA-7245: Deprecate WindowStore#put(key, value) (#7105)
  KAFKA-8179: Part 7, cooperative rebalancing in Streams (#7386)
  KAFKA-8985; Add flexible version support to inter-broker APIs (#7453)
  MINOR: Bump version to 2.5.0-SNAPSHOT (#7455)

ijuma committed:
WIP

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk: (410 commits)
  KAFKA-8843: KIP-515: Zookeeper TLS support
  MINOR: Add missing quote for malformed line content (#8070)
  MINOR: Simplify KafkaProducerTest (#8044)
  KAFKA-9507; AdminClient should check for missing committed offsets (#8057)
  KAFKA-9519: Deprecate the --zookeeper flag in ConfigCommand (#8056)
  KAFKA-9509; Fixing flakiness of MirrorConnectorsIntegrationTest.testReplication (#8048)
  HOTFIX: Fix two test failures in JDK11 (#8063)
  DOCS - clarify transactionalID and idempotent behavior (#7821)
  MINOR: further InternalTopologyBuilder cleanup  (#8046)
  MINOR: Add timer for update limit offsets (#8047)
  HOTFIX: Fix spotsbug failure in Kafka examples (#8051)
  KAFKA-9447: Add new customized EOS model example (#8031)
  KAFKA-8164: Add support for retrying failed (#8019)
  HOTFIX: checkstyle for newly added unit test
  KAFKA-9261; Client should handle unavailable leader metadata (#7770)
  MINOR: Fix typos introduced in KIP-559 (#8042)
  MINOR: Fixing null handilg in ValueAndTimestampSerializer (#7679)
  KAFKA-9113: Clean up task management and state management (#7997)
  MINOR: fix checkstyle issue in ConsumerConfig.java (#8038)
  KAFKA-9491; Increment high watermark after full log truncation (#8037)
  ...

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk: (366 commits)
  MINOR: Improve producer test BufferPoolTest#testCloseNotifyWaiters. (#7982)
  MINOR: document how to escape json parameters to ducktape tests (#8546)
  KAFKA-9885; Evict last members of a group when the maximum allowed is reached (#8525)
  KAFKA-9866: Avoid election for topics where preferred leader is not in ISR (#8524)
  KAFKA-9839; Broker should accept control requests with newer broker epoch (#8509)
  KAKFA-9612: Add an option to kafka-configs.sh to add configs from a prop file (KIP-574)
  MINOR: Partition is under reassignment when adding and removing (#8364)
  MINOR: reduce allocations in log start and recovery checkpoints (#8467)
  MINOR: Remove unused foreign-key join class (#8547)
  HOTFIX: Fix broker bounce system tests (#8532)
  KAFKA-9704: Fix the issue z/OS won't let us resize file when mmap. (#8224)
  KAFKA-8639: Replace AddPartitionsToTxn with Automated Protocol  (#8326)
  MINOR: equals() should compare all fields for generated classes (#8539)
  KAFKA-9844; Fix race condition which allows more than maximum number of members(#8454)
  KAFKA-9823: Remember the sent generation for the coordinator request (#8445)
  KAFKA-9883: Add better error message when REST API forwards a request and leader is not known (#8536)
  KAFKA-9907: Switch default build to Scala 2.13 (#8537)
  MINOR: Some html fixes in Streams DSL documentation (#8503)
  MINOR: Enable fatal warnings with scala 2.13 (#8429)
  KAFKA-9852: Change the max duration that calls to the buffer pool can block from 2000ms to 10ms to reduce overall test runtime (#8464)
  ...

ijuma committed:
WIP

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk: (248 commits)
  KAFKA-10049: Fixed FKJ bug where wrapped serdes are set incorrectly when using default StreamsConfig serdes (#8764)
  KAFKA-10027: Implement read path for feature versioning system (KIP-584) (#8680)
  KAFKA-10085: correctly compute lag for optimized source changelogs (#8787)
  KAFKA-10086: Integration test for ensuring warmups are effective (#8818)
  KAFKA-9374: Make connector interactions asynchronous (#8069)
  MINOR: reduce sizeInBytes for percentiles metrics (#8835)
  KAFKA-10115: Incorporate errors.tolerance with the Errant Record Reporter (#8829)
  KAFKA-9216: Enforce that Connect’s internal topics use `compact` cleanup policy (#8828)
  KAFKA-9845: Warn users about using config providers with plugin.path property (#8455)
  KAFKA-7833: Add missing test (#8847)
  KAFKA-9066: Retain metrics for failed tasks (#8502)
  KAFKA-9841: Revoke duplicate connectors and tasks when zombie workers return with an outdated assignment (#8453)
  KAFKA-9985: Sink connector may exhaust broker when writing in DLQ (#8663)
  KAFKA-9441: remove prepareClose() to simplify task management (#8833)
  KAFKA-7833: Add Global/StateStore name conflict check (#8825)
  KAFKA-9969: Exclude ConnectorClientConfigRequest from class loading isolation (#8630)
  KAFKA-9991: Fix flaky unit tests (#8843)
  KAFKA-10014; Always try to close all channels in Selector#close (#8685)
  KAFKA-10079: improve thread-level stickiness (#8775)
  MINOR: Print all removed dynamic members during join complete (#8816)
  ...

ijuma committed:
WIP

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk: (148 commits)
  MINOR: remove NewTopic#NO_PARTITIONS and NewTopic#NO_REPLICATION_FACTOR as they are duplicate to CreateTopicsRequest#NO_NUM_PARTITIONS and CreateTopicsRequest#NO_REPLICATION_FACTOR (#9077)
  MINOR: Remove staticmethod tag to be able to use logger of instance (#9086)
  MINOR: Adjust 'release.py' script to use shell when using gradlewAll and PGP signing, which were required to build the 2.6.0 RCs (#9045)
  MINOR: Update dependencies for Kafka 2.7 (part 1) (#9082)
  MINOR: INFO log4j when request re-join (#9068)
  MINOR: Recommend Java 11 (#9080)
  KAFKA-10306: GlobalThread should fail on InvalidOffsetException (#9075)
  KAFKA-10158: Fix flaky testDescribeUnderReplicatedPartitionsWhenReassignmentIsInProgress (#9022)
  MINOR: code cleanup for `VOut` inconsistent naming (#8907)
  KAFKA-10246 : AbstractProcessorContext topic() throws NPE (#9034)
  KAFKA-10305: Print usage when parsing fails for ConsumerPerformance (#9071)
  MINOR: removed incorrect deprecation annotations (#9061)
  MINOR: speed up release script (#9070)
  MINOR: add task ':streams:testAll' (#9073)
  KAFKA-10301: Do not clear Partition#remoteReplicasMap during partition assignment updates (#9065)
  KAFKA-10268: dynamic config like "--delete-config log.retention.ms" doesn't work (#9051)
  KAFKA-10300 fix flaky core/group_mode_transactions_test.py (#9059)
  MINOR: Publish metrics package in the javadoc (#9036)
  KAFKA-8264: decrease the record size for flaky test
  KAFKA-5876: Add new exception types for Interactive Queries (#8200)
  ...

chia7712 commented:
@ijuma Is this still active? All requests/responses are using auto-generated protocol so it is time to get rid of `Struct` from serialization. If you are busy with other works, I have free cycle to address this issue :)

ijuma committed:
Fix merge

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk: (405 commits)
KAFKA-6687: restrict DSL to allow only Streams from the same source
topics (#9609)
  MINOR: Small cleanups in `AlterIsr` handling logic (#9663)
MINOR: Increase unit test coverage of method
ProcessorTopology#updateSourceTopics() (#9654)
  MINOR: fix reading SSH output in Streams system tests (#9665)
  KAFKA-10770: Remove duplicate defination of Metrics#getTags (#9659)
  KAFKA-10722: Described the types of the used state stores (#9607)
  KAFKA-10702; Skip bookkeeping of empty transactions (#9632)
  MINOR: Remove erroneous extra <code> in design doc (#9657)
KAFKA-10736 Convert transaction coordinator metadata schemas to use g…
(#9611)
  MINOR: Update vagrant/tests readme (#9650)
  KAFKA-10720: Document prohibition on header mutation by SMTs (#9597)
  KAFKA-10713: Stricter protocol parsing in hostnames (#9593)
  KAFKA-10565: Only print console producer prompt with a tty (#9644)
  MINOR: fix listeners doc to close <code> properly (#9655)
  MINOR: Remove unnecessary statement from WorkerConnector#doRun (#9653)
KAFKA-10758: ProcessorTopology should only consider its own nodes when
updating regex source topics (#9648)
KAFKA-10754: fix flaky tests by waiting kafka streams be in running
state before assert (#9629)
  MINOR: Upgrade to Scala 2.13.4 (#9643)
  MINOR: Update build and test dependencies (#9645)
MINOR: Remove redundant argument from TaskMetricsGroup#recordCommit
(#9642)
  ...

clients/src/main/java/org/apache/kafka/clients/ClientRequest.java
clients/src/main/java/org/apache/kafka/clients/NetworkClient.java
clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java
clients/src/main/java/org/apache/kafka/common/requests/AbstractRequestResponse.java
clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java
clients/src/main/java/org/apache/kafka/common/requests/AlterConfigsResponse.java
clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java
clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java
clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java
clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrRequest.java
clients/src/main/java/org/apache/kafka/common/requests/ListOffsetRequest.java
clients/src/main/java/org/apache/kafka/common/requests/ListOffsetResponse.java
clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochRequest.java
clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochResponse.java
clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java
clients/src/main/java/org/apache/kafka/common/requests/ProduceResponse.java
clients/src/main/java/org/apache/kafka/common/requests/RequestHeader.java
clients/src/main/java/org/apache/kafka/common/requests/StopReplicaRequest.java
clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java
clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java
clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java
clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java
clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java
clients/src/test/java/org/apache/kafka/common/requests/RequestContextTest.java
clients/src/test/java/org/apache/kafka/common/requests/RequestHeaderTest.java
clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java

ijuma committed:
Minor style fixes

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk:
KAFKA-9263 The new hw is added to incorrect log when
ReplicaAlterLogDirsThread is replacing log (fix
PlaintextAdminIntegrationTest.testAlterReplicaLogDirs) (#9423)
  KAFKA-10729; Bump remaining RPC's to use tagged fields. (#9601)

clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java
clients/src/main/java/org/apache/kafka/common/requests/AlterReplicaLogDirsResponse.java
clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java

ijuma committed:
Complete reimplementation of WriteTxnMarkersResponse

ijuma committed:
AbstractResponse toString no longer requires version

ijuma committed:
Various minor fixes

ijuma committed:
Complete implementation of AlterClientQuotasRequest/Response

ijuma committed:
Minor style fixes

ijuma committed:
Revert SaslAuthenticateRequest.Builder.toString change as it could leak authentication data

ijuma committed:
Fix bug introduced in previous commit where handleProduceRequest did not account for header bytes

ijuma committed:
Minor style fixes

ijuma committed:
Complete implementation of DescribeConfigsResponse

ijuma committed:
Fix existing bug in ProduceRequest uncovered by AbstractResponse.toString

The test for this functionality was broken before the version parameter
was removed from AbstractResponse.toString.

ijuma committed:
Style fix

ijuma committed:
Fix NPEs in DeleteAclsResponse and DescribeAclsResponse

ijuma committed:
RequestResponseTest fixes

ijuma committed:
Fix NetworkClientTest failures

ijuma committed:
Fix KafkaApisTest failures and rename errors to errorsByProducerId for clarity

ijuma committed:
Minor style fix

ijuma committed:
Fix RequestResponseTest and error message in DescribeClientQuotasRequest

ijuma committed:
Fix DeleteAclsResponseTest

ijuma committed:
Fix EndTxnResponseTest

ijuma committed:
Remove TestUtils.serializeMessage in favor of MessageTestUtil.messageToByteBuffer

ijuma committed:
Fix LeaveGroupResponseTest

ijuma committed:
Minor fixes

ijuma committed:
Test fixes/improvements

ijuma committed:
Merge remote-tracking branch 'apache-github/trunk' into skip-to-struct-for-generated-requests

* apache-github/trunk:
MINOR: Fix flaky test shouldQueryOnlyActivePartitionStoresByDefault
(#9681)
  KAFKA-10799 AlterIsr utilizes ReplicaManager ISR metrics (#9677)
  MINOR: Fix KTable-KTable foreign-key join example (#9683)
KAFKA-10473: Add docs on partition size-on-disk, and other log-related
metrics (#9276)
  KAFKA-10739; Replace EpochEndOffset with automated protocol (#9630)
KAFKA-10460: ReplicaListValidator format checking is incomplete
(#9326)
KAFKA-10554; Perform follower truncation based on diverging epochs in
Fetch response (#9382)
  MINOR: Align the UID inside/outside container (#9652)
KAFKA-10794 Replica leader election is too slow in the case of too
many partitions (#9675)
KAFKA-10090 Misleading warnings: The configuration was supplied but i…
(#8826)

clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochResponse.java
clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java
core/src/test/scala/unit/kafka/server/epoch/util/ReplicaFetcherMockBlockingSend.scala

ijuma committed:
Fix MetadataRequestTest failure

ijuma committed:
Use SendBuilder in AbstractRequest/Response

Also consolidate and clean-up.

ijuma committed:
Fix request size returned by parseRequest

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractResponse.java:
```diff
@@ -84,148 +95,151 @@ protected void updateErrorCounts(Map<Errors, Integer> errorCounts, Errors error)
+        AbstractResponse response = AbstractResponse.parseResponse(apiKey, buffer, apiVersion);
+
+        // We correlate after parsing the response to avoid spurious correlation errors when receiving malformed
+        // responses
```
TODO: Handle this better.

ijuma committed:
Revert correlation id verification order change

ijuma committed:
Revert MetadataResponse.json change

ijuma commented:
Looks like the failures are unrelated, so this is ready for review @chia7712. The PR description needs a few tweaks, I will do that tomorrow.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java:
```diff
@@ -95,21 +93,33 @@ public short version() {
+    }
+
+    // Visible for testing
+    public final ByteBuffer serializeWithHeader(RequestHeader header) {
```
How about moving this method to test code? ```MessageTestUtil``` could be a good place.

chia7712 replied:
ditto for following "Visible for testing" code

chia7712 replied:
BTW, this method is duplicate to TestUtils#serializeRequestHeader

ijuma replied:
This method serializes the request with the header while the test method serializes the header only. So, not a duplicate, right? Personally, I think it's useful for the class to be symmetric: it can parse from a buffer and serialize to a buffer. Do you have a concern that it may be misused?

ijuma replied:
There is a reasonable argument that `serializeBody` is the counterpart to `parse` though and this particular method doesn't have to be here. I can buy that.

chia7712 replied:
> Do you have a concern that it may be misused?

yes. Also the various serialization methods are a bit chaos to me :(

chia7712 replied:
Maybe we should move all testing-only serialization to a single utils (in test scope). I grep "static ByteBuffer" + "request/response" and ```MessageUtils```, ```MessageTestUtil```, ```RequestUtils``` and ```TestUtils``` have similar method interface.

ijuma replied:
I have a refactoring that consolidates the logic around `Send`. But I think that would be best done in a separate PR since this PR is really massive. Since these methods are all in internal classes and are only used by tests, I suggest we keep them as they are for now and tackle these nits in the follow up PR.

ijuma replied:
I have these changes locally, so I can submit them immediately after.

ijuma commented:
Actually, I went ahead and updated the PR description.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java:
```diff
@@ -95,21 +93,33 @@ public short version() {
 
     public String toString(boolean verbose) {
-        return toStruct().toString();
+        return data().toString();
```
Is this toString variety still useful?

ijuma replied:
I think so. Do you think it's not?

ijuma replied:
Sorry, what do you mean by `variety`? The `verbose` boolean? Yes, that is useful for things like fetch and produce.

chia7712 replied:
It seems only ```ProduceRequest``` does produce verbose string. Maybe we can remove the ```toString(boolean verbose)``` as most requests does not generate verbose string. Also, ```ProduceRequest``` should override ```toString()``` to generate a verbose string always.

ijuma replied:
The verbosity flag is used for request logging. Produce and fetch are the most important ones. The rest can be ignored. I don't see the benefit in removing this functionality.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsRequest.java:
```diff
@@ -96,8 +92,8 @@ public boolean isValid() {
     @Override
-    protected Struct toStruct() {
-        return data.toStruct(version());
+    public ApiVersionsRequestData data() {
```
```public``` is for testing?

ijuma replied:
No, SaslServerAuthenticator uses this.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/protocol/SendBuilder.java:
```diff
@@ -209,16 +210,13 @@ public static Send buildResponseSend(
+        Message apiMessage,
         short apiVersion
     ) {
         ObjectSerializationCache serializationCache = new ObjectSerializationCache();
```
not sure whether we should reduce the initial capacity of ObjectSerializationCache. Not all request/response uses full cache.

ijuma replied:
Maybe we can file a JIRA and investigate this separately. Not clear what would be the optimal size.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java:
```diff
@@ -90,4 +75,37 @@ public static boolean hasTransactionalRecords(ProduceRequest request) {
+        ObjectSerializationCache serializationCache = new ObjectSerializationCache();
+        MessageSizeAccumulator messageSize = RequestUtils.size(serializationCache, header, headerVersion, apiMessage, apiVersion);
+
+        ByteBuffer buffer = ByteBuffer.allocate(messageSize.totalSize());
```
Should it subtract ```messageSize.zeroCopySize()```?

ijuma replied:
Yeah, good point. I don't think it affects its current non test usage, but could result in a bug as the code changes. Will fix.

ijuma replied:
Actually, I need to think about this more. Because `SendBuilder` is not being used, we actually need to copy all the bytes to the returned buffer. I need to check what happens for a message contains records.

chia7712 replied:
oh, you are right. It uses ```ByteBufferAccessor``` rather than ```SendBuilder```.  There is no zero copy.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/requests/RequestHeader.java:
```diff
@@ -81,14 +70,23 @@ public RequestHeaderData data() {
         return data;
     }
 
+    public void write(ByteBuffer buffer, ObjectSerializationCache serializationCache) {
```
This is used by testing only. Maybe we can move it to test scope.

chia7712 reviewed (commented):
@ijuma Thanks for this huge patch. Some minor comments are left. Please take a look.

chia7712 reviewed (approved):
+1 to this great cleanup and improvement. Please merge it to trunk and then we can address the TODOs in separate PRs :)

ijuma commented:
Thanks for the review! Will post the follow up in a bit.

dajac commented:
Nice one @ijuma!

ijuma commented:
Follow up PR submitted https://github.com/apache/kafka/pull/9714
